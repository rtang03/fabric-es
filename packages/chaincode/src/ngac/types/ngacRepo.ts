import { Attribute } from './attribute';
import { Policy } from './policy';
import { Resource } from './resource';

export interface NgacRepo {
  // add one policy
  addPolicy: (policy: Policy) => Promise<Policy>;
  // add and replace one attribute group to a MSPID
  addMSPAttr: (resource: Resource) => Promise<Attribute[]>;
  // add and replace one attribute group to a reource URI
  addResourceAttr: (resource: Resource) => Promise<Attribute[]>;
  deleteMSPAttrByMSPID: (mspid: string) => Promise<string>;
  // delete all policies by id
  deletePolicyById: (id: string) => Promise<string[]>;
  // delete one policy by id
  deletePolicyByIdSid: (id: string, sid: string) => Promise<string>;
  // delete attribute group by URI
  deleteReourceAttrByURI: (uri: string) => Promise<string>;
  // delete attribute group by MSPID
  getMSPAttrByMSPID: (mspid: string) => Promise<Attribute[]>;
  // get policies by id
  getPolicyById: (id: string) => Promise<Policy[]>;
  // get one policy
  getPolicyByIdSid: (id: string, sid: string) => Promise<Policy>;
  // get array of attribute group by URI path segment
  // this API may be useless; no such use case. This is not exposed as chaincode interface
  getResourceAttrGroupByURI: (uri: string) => Promise<Attribute[][]>;
  // get attribte gropu by URI
  getResourceAttrByURI: (uri: string) => Promise<Attribute[]>;
  // update and insert the attribute group, by appending the existing one
  upsertResourceAttr: (resource: Resource) => Promise<Attribute[]>;
}
