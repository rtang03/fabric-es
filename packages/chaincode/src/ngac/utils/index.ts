export * from './makeKey';
export * from './createResource';
export * from './createPolicy';
export * from './createId';
export * from './hasList';
export * from './stringEquals';
