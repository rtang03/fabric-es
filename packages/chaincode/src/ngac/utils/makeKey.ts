export const makeKey = (keyParts: string[]) => keyParts.map(part => JSON.stringify(part)).join('~');

export const splitKey = (key: string) => key.split('~');
