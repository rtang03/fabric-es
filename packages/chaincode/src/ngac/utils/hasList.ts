import { includes, intersection } from 'lodash';
import { Assertion, Attribute } from '../types';

const allowOrDeny = { Allow: true, Deny: false };

export const hasList: (option: {
  sid: string;
  effect: string;
  requirement: Attribute[];
  resourceAttrs: Attribute[];
  condition: any;
  debug?: boolean;
}) => Assertion[] = ({ sid, effect, resourceAttrs, requirement, condition, debug = false }) =>
  !condition.hasList
    ? [{ sid, assertion: true, message: 'No hasList condition defined' }]
    : Object.entries<string>(condition.hasList).map(([actionToDo, authorizedAttribute]) => {
        const requiredAttribute = requirement.find(({ key }) => key === actionToDo);
        const assertion = resourceAttrs.reduce((prev, { type, key, value }) => {
          if (!requiredAttribute) return false;
          const result =
            prev ||
            (key !== authorizedAttribute
              ? false
              : type === '1'
              ? includes(requiredAttribute.value, value)
              : !!intersection(requiredAttribute.value, value).length);
          if (debug)
            console.log(
              `Target-attr=${key}/${value} vs Authorized-attr=${authorizedAttribute}  -> fulfilling ${JSON.stringify(
                requiredAttribute
              )} -> result=${result}`
            );
          return result;
        }, false);

        return assertion
          ? {
              sid,
              assertion: allowOrDeny[effect]
            }
          : {
              sid,
              assertion: !allowOrDeny[effect]
            };
      });
