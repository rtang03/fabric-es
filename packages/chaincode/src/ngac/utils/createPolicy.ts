import { Context } from 'fabric-contract-api';
import { Policy } from '../types';
import { createId } from './createId';
import { makeKey } from './makeKey';

export const createPolicy: (option: {
  context: Context;
  policyClass?: string;
  sid: string;
  allowedEvents: string[];
  uri: string;
  condition?: {
    hasList?: any;
    stringEquals?: any;
  };
  effect?: string;
}) => Policy = ({ context, policyClass = 'default', sid, allowedEvents, uri, condition, effect = 'Allow' }) => {
  const x509id = createId([
    context.clientIdentity.getMSPID(),
    context.clientIdentity.getX509Certificate().subject.commonName
  ]);
  const key = makeKey([x509id, sid]);

  if (!x509id || !sid || !uri || !context || !allowedEvents || !allowedEvents.length) {
    return null;
  }

  const basePolicy = {
    key,
    policyClass,
    sid,
    allowedEvents,
    attributes: {
      uri
    },
    effect
  };

  if (condition) (basePolicy as any).condition = condition;

  return basePolicy;
};
