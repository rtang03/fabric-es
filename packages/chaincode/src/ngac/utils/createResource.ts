import { Context } from 'fabric-contract-api';
import { Attribute, CONTEXT, Resource } from '../types';
import { createId } from './createId';
import { makeKey } from './makeKey';

export const createResource: (option: {
  context: Context;
  mspAttrs?: Attribute[];
  entityName: string;
  entityId?: string;
  resourceAttrs?: Attribute[];
}) => Resource = ({ context, mspAttrs, entityName, entityId, resourceAttrs }) => {
  if (!entityName || !context || !resourceAttrs) {
    return null;
  }
  const cid = context.clientIdentity;
  const contextAttrs: Attribute[] = [];

  // todo: replace with optional chaining to avoid TypeError: cannot read property of undefined
  const invoker_id = createId([cid.getMSPID(), cid.getX509Certificate().subject.commonName]);
  const invoker_mspid = cid.getMSPID();
  const subject_cn = cid.getX509Certificate().subject.commonName;
  const subject_orgname = cid.getX509Certificate().subject.organizationName;
  const subject_ouname = cid.getX509Certificate().subject.organizationalUnitName;
  const issuer_cn = cid.getX509Certificate().issuer.commonName;
  const issuer_orgname = cid.getX509Certificate().issuer.organizationName;
  const attr = (key, value) => (value ? { type: '1' as any, key, value } : null);

  contextAttrs.push(
    ...[
      attr(CONTEXT.INVOKER_MSPID, invoker_mspid),
      attr(CONTEXT.INVOKER_ID, invoker_id),
      attr(CONTEXT.SUBJECT_CN, subject_cn),
      attr(CONTEXT.SUBJECT_ORGNAME, subject_orgname),
      attr(CONTEXT.SUBJECT_OUNAME, subject_ouname),
      attr(CONTEXT.ISSUER_CN, issuer_cn),
      attr(CONTEXT.ISSUER_ORGNAME, issuer_orgname)
    ].filter(item => !!item)
  );

  return entityId
    ? {
        key: makeKey(['model', invoker_mspid, entityName, entityId]),
        uri: `model/${invoker_mspid}/${entityName}/${entityId}`,
        contextAttrs,
        mspAttrs: mspAttrs || [],
        resourceAttrs: resourceAttrs || []
      }
    : {
        key: makeKey(['model', invoker_mspid, entityName]),
        uri: `model/${invoker_mspid}/${entityName}`,
        contextAttrs,
        mspAttrs: mspAttrs || [],
        resourceAttrs: resourceAttrs || []
      };
};
