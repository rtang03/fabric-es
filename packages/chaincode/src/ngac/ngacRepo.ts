import { Context } from 'fabric-contract-api';
import { assign, values } from 'lodash';
import stateList from './ledger-api/statelist';
import { Attribute, NAMESPACE as NS, NgacRepo, Policy } from './types';
import { makeKey } from './utils';

export const ngacRepo: (context: Context) => NgacRepo = context => ({
  addMSPAttr: async resource =>
    resource
      ? stateList<Attribute[]>(NS.MSP_ATTRIBUTE, context).addState(
          resource.key,
          resource.mspAttrs
        )
      : null,
  addResourceAttr: async resource =>
    resource
      ? stateList<Attribute[]>(NS.RESOURCE_ATTRIBUTE, context).addState(
          resource.key,
          resource.resourceAttrs
        )
      : null,
  upsertResourceAttr: async resource => {
    if (!resource) return null;
    const attributes = await stateList<Attribute[]>(
      NS.RESOURCE_ATTRIBUTE,
      context
    ).getState(resource.key);

    const newAttrs = {};
    resource.resourceAttrs.forEach(attr => (newAttrs[attr.key] = attr));
    const existingAttrs = {};
    attributes.forEach(attr => (existingAttrs[attr.key] = attr));
    const resultAttrs = values<Attribute>(assign({}, existingAttrs, newAttrs));

    return resultAttrs
      ? stateList<Attribute[]>(NS.RESOURCE_ATTRIBUTE, context).addState(
          resource.key,
          resultAttrs.map(attr => {
            if (attr.key === 'version') {
              let version = parseInt(attr.value as string, 10);
              return {
                type: '1',
                key: 'version',
                value: `${++version}`,
                immutable: false
              };
            }
            return attr;
          })
        )
      : null;
  },
  addPolicy: async policy =>
    policy
      ? stateList<Policy>(NS.POLICY, context).addState(policy.key, policy)
      : null,
  deleteMSPAttrByMSPID: async mspid =>
    mspid
      ? stateList(NS.MSP_ATTRIBUTE, context).deleteStateByKey(makeKey([mspid]))
      : null,
  deletePolicyById: async id =>
    id
      ? stateList(NS.POLICY, context).deleteStatesByKeyRange([
          JSON.stringify(id)
        ])
      : null,
  deletePolicyByIdSid: async (id, sid) =>
    id && sid
      ? stateList(NS.POLICY, context).deleteStateByKey(makeKey([id, sid]))
      : null,
  deleteReourceAttrByURI: async uri =>
    uri
      ? stateList(NS.RESOURCE_ATTRIBUTE, context).deleteStateByKey(
          makeKey(uri.split('/'))
        )
      : null,
  getMSPAttrByMSPID: async mspid =>
    mspid
      ? stateList<Attribute[]>(NS.MSP_ATTRIBUTE, context).getState(
          makeKey([mspid])
        )
      : null,
  getPolicyById: async id =>
    id
      ? stateList<Policy>(NS.POLICY, context).getQueryResult([
          JSON.stringify(id)
        ])
      : null,
  getPolicyByIdSid: async (id, sid) =>
    id && sid
      ? stateList<Policy>(NS.POLICY, context).getState(makeKey([id, sid]))
      : null,
  getResourceAttrGroupByURI: async uri =>
    uri
      ? stateList<Attribute[]>(NS.RESOURCE_ATTRIBUTE, context).getQueryResult(
          uri.split('/').map(part => JSON.stringify(part))
        )
      : null,
  getResourceAttrByURI: async uri =>
    uri
      ? stateList<Attribute[]>(NS.RESOURCE_ATTRIBUTE, context).getState(
          makeKey(uri.split('/'))
        )
      : null
});
