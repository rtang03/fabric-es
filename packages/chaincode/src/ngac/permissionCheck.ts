import { Context } from 'fabric-contract-api';
import { isEqual } from 'lodash';
import { ngacRepo } from './ngacRepo';
import { policyDecisionEngine } from './policyDecisionEngine';
import { postAssertion } from './postAssertion';
import { Assertion, Attribute, CONTEXT, NAMESPACE, RESOURCE } from './types';
import { createId, createResource } from './utils';

enum FCN {
  CREATE_COMMIT = 'createCommit',
  QUERY_BY_ENTITYID = 'queryByEntityId',
  QUERY_BY_ENTITYID_COMMITID = 'queryByEntityIdCommitId',
  DELETE_BY_ENTITYID_COMMITID = 'deleteByEntityIdCommitId',
  DELETE_BY_ENTITYID = 'deleteByEntityId',
  ADD_MSP_ATTR = 'addMSPAttr',
  ADD_RESORUCE_ATTR = 'addResourceAttr',
  ADD_POLICY = 'addPolicy',
  DELETE_MSP_ATTR_BY_MSPID = 'deleteMSPAttrByMSPID',
  DELETE_POLICY_BY_ID = 'deletePolicyById',
  DELETE_POLICY_BY_ID_SID = 'deletePolicyByIdSid',
  DELETE_RESOURCE_ATTR_BY_URI = 'deleteReourceAttrByURI',
  UPSERT_RESOURCE_ATTR = 'upsertResourceAttr',
  GET_MSP_ATTR_BY_MSPID = 'getMSPAttrByMSPID',
  GET_POLICY_BY_ID = 'getPolicyById',
  GET_POLICY_BY_ID_SID = 'getPolicyByIdSid',
  GET_RESOURCE_ATTR_BY_URI = 'getResourceAttrByURI'
}

const noPolicyRequired = Promise.resolve([
  { sid: 'system', message: 'No policy required', assertion: true }
]);

export const permissionCheck: (option: {
  policyClass?: string;
  context: Context;
}) => Promise<Assertion[]> = async ({ context, policyClass }) => {
  const { stub, clientIdentity } = context;
  const { fcn, params } = stub.getFunctionAndParameters();
  const type: any = '1';
  const attr = (key, value, immutable = true) =>
    value ? { type, key, value, immutable } : null;
  const mspAttrs = await ngacRepo(context).getMSPAttrByMSPID(
    clientIdentity.getMSPID()
  );
  let resourceAttrs: Attribute[] = [];
  const cn = clientIdentity.getX509Certificate().subject.commonName;
  const mspid = clientIdentity.getMSPID();
  const id = createId([mspid, cn]);
  return {
    [FCN.CREATE_COMMIT]: async (): Promise<Assertion[]> => {
      const [entityName, entityId, version, evenStr] = params;
      resourceAttrs = [
        attr(CONTEXT.INVOKER_ID, id),
        attr(CONTEXT.INVOKER_MSPID, mspid),
        attr(CONTEXT.SUBJECT_CN, cn)
      ].filter(item => !!item);

      if (version === '0') {
        [
          attr(RESOURCE.VERSION, version, false),
          attr(RESOURCE.ENTITYNAME, entityName),
          attr(RESOURCE.ENTITYID, entityId),
          attr(RESOURCE.CREATOR_MSPID, mspid),
          attr(RESOURCE.CREATOR_CN, cn),
          attr(RESOURCE.CREATOR_ID, id)
        ]
          .filter(item => !!item)
          .forEach(item => resourceAttrs.push(item));
      } else {
        const uri = `${NAMESPACE.MODEL}/${mspid}/${entityName}/${entityId}`;
        const attrs = await ngacRepo(context).getResourceAttrByURI(uri);

        if (isEqual(attrs, [])) {
          return [
            { sid: 'system', assertion: false, message: 'No resource found' }
          ];
        } else attrs.forEach(item => resourceAttrs.push(item));
      }

      const eventsJson: any[] = JSON.parse(evenStr);
      const eventTypes = eventsJson.map(({ type }) => type);
      const policies = await ngacRepo(context).getPolicyById(id);
      const target = createResource({
        context,
        entityName,
        entityId,
        resourceAttrs,
        mspAttrs
      });
      return isEqual(policies, [])
        ? [{ sid: 'system', assertion: false, message: 'No policy found' }]
        : policyDecisionEngine(policies, context)
            .request({ eventTypes, target })
            .then(async assertions => {
              await postAssertion(assertions, target, context);
              return assertions;
            });
    },
    [FCN.QUERY_BY_ENTITYID]: () => noPolicyRequired,
    [FCN.QUERY_BY_ENTITYID_COMMITID]: () => noPolicyRequired,
    [FCN.DELETE_BY_ENTITYID]: () => noPolicyRequired,
    [FCN.DELETE_BY_ENTITYID_COMMITID]: () => noPolicyRequired,
    [FCN.ADD_MSP_ATTR]: () => noPolicyRequired,
    [FCN.ADD_RESORUCE_ATTR]: () => noPolicyRequired,
    [FCN.ADD_POLICY]: () => noPolicyRequired,
    [FCN.DELETE_POLICY_BY_ID]: () => noPolicyRequired,
    [FCN.DELETE_POLICY_BY_ID_SID]: () => noPolicyRequired,
    [FCN.DELETE_RESOURCE_ATTR_BY_URI]: () => noPolicyRequired,
    [FCN.DELETE_MSP_ATTR_BY_MSPID]: () => noPolicyRequired,
    [FCN.GET_POLICY_BY_ID]: () => noPolicyRequired,
    [FCN.GET_POLICY_BY_ID_SID]: () => noPolicyRequired,
    [FCN.GET_RESOURCE_ATTR_BY_URI]: () => noPolicyRequired,
    [FCN.GET_MSP_ATTR_BY_MSPID]: () => noPolicyRequired,
    [FCN.UPSERT_RESOURCE_ATTR]: () => noPolicyRequired
  }[fcn]() as Promise<Assertion[]>;
};
