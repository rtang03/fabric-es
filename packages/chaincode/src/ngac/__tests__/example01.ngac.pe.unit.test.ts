import { ngacRepo } from '../ngacRepo';
import { permissionCheck } from '../permissionCheck';
import { createId } from '../utils';
import { mspAttributeDb, policyDb, resourceAttributeDb } from './__utils__';

jest.mock('../ledger-api/statelist');

const context: any = {
  stub: {
    getFunctionAndParameters: jest.fn()
  },
  clientIdentity: {
    getMSPID: jest.fn(),
    getID: jest.fn(),
    getX509Certificate: jest.fn()
  },
  resourceAttributeDb,
  mspAttributeDb,
  policyDb
};
let entityName;
let entityId;
let version;
let eventStr;
let id;

describe('Example 1: PolicyEngine Tests', () => {
  beforeEach(() => {
    context.clientIdentity.getMSPID.mockImplementation(() => 'Org1MSP');
    context.clientIdentity.getX509Certificate.mockImplementation(() => ({
      subject: { commonName: 'Admin@example.com' },
      issuer: { commonName: 'rca-org1' }
    }));
    entityName = 'dev_ngac_example1';
    entityId = 'ngac_unit_01';
    version = '0';
    eventStr = JSON.stringify([{ type: 'DocumentCreated' }]);
    id = createId(['Org1MSP', 'Admin@example.com']);
    context.stub.getFunctionAndParameters.mockImplementation(() => ({
      fcn: 'createCommit',
      params: [entityName, entityId, version, eventStr]
    }));
    context.clientIdentity.getID.mockImplementation(() => id);
  });

  // policy found and asserted
  it('1a: should createDocument', async () => {
    await permissionCheck({ context }).then(assertions =>
      expect(assertions).toEqual([
        { sid: 'allowCreateDocument', assertion: true }
      ])
    );
    await ngacRepo(context)
      .getResourceAttrGroupByURI('model/Org1MSP/dev_ngac_example1/ngac_unit_01')
      .then(attr => expect(attr).toMatchSnapshot());
  });

  // policy not found at all
  it('1b: should fail updateDocument, which is not allowedEvents', async () => {
    context.stub.getFunctionAndParameters.mockImplementationOnce(() => ({
      fcn: 'createCommit',
      params: [
        entityName,
        entityId,
        version,
        JSON.stringify([{ type: 'DocumentUpdated' }])
      ]
    }));
    return permissionCheck({ context }).then(assertions =>
      expect(assertions).toEqual([
        { sid: 'system', assertion: false, message: 'No policy found' }
      ])
    );
  });

  // policy found, but assert false
  it('1c: should fail createDocument with wrong ID', async () => {
    context.clientIdentity.getX509Certificate.mockImplementation(() => ({
      subject: { commonName: 'wrong@example.com' },
      issuer: { commonName: 'rca-org1' }
    }));
    return permissionCheck({ context }).then(assertions =>
      expect(assertions).toEqual([
        { sid: 'system', message: 'No policy found', assertion: false }
      ])
    );
  });

  // Policy found, but missing resource attributes cannot proceed assertion
  it('1e: should fail getResourceAttr', async () => {
    context.stub.getFunctionAndParameters.mockImplementationOnce(() => ({
      fcn: 'createCommit',
      params: ['non-existing entityName', entityId, version, eventStr]
    }));
    return permissionCheck({ context }).then(assertions =>
      expect(assertions).toEqual([
        {
          sid: 'allowCreateDocument',
          assertion: false,
          message: 'Cannot find resource attributes'
        }
      ])
    );
  });

  // The Invoker's mspid is Org2MSP, cannot find the policy
  it('1f: should fail createDocument with wrong mspid', async () => {
    context.clientIdentity.getMSPID.mockImplementation(() => 'Org2MSP');
    return permissionCheck({ context }).then(assertions =>
      expect(assertions).toEqual([
        {
          sid: 'system',
          assertion: false,
          message: 'No policy found'
        }
      ])
    );
  });
});
