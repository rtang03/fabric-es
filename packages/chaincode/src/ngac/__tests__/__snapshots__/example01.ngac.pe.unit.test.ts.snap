// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example 1: PolicyEngine Tests 1a: should createDocument 1`] = `
Array [
  Object {
    "immutable": true,
    "key": "invoker_id",
    "type": "1",
    "value": "Org1MSP::Admin@example.com",
  },
  Object {
    "immutable": true,
    "key": "invoker_mspid",
    "type": "1",
    "value": "Org1MSP",
  },
  Object {
    "immutable": true,
    "key": "subject_cn",
    "type": "1",
    "value": "Admin@example.com",
  },
  Object {
    "immutable": false,
    "key": "version",
    "type": "1",
    "value": "0",
  },
  Object {
    "immutable": true,
    "key": "entityName",
    "type": "1",
    "value": "dev_ngac_example1",
  },
  Object {
    "immutable": true,
    "key": "entityId",
    "type": "1",
    "value": "ngac_unit_01",
  },
  Object {
    "immutable": true,
    "key": "creator_mspid",
    "type": "1",
    "value": "Org1MSP",
  },
  Object {
    "immutable": true,
    "key": "creator_cn",
    "type": "1",
    "value": "Admin@example.com",
  },
  Object {
    "immutable": true,
    "key": "creator_id",
    "type": "1",
    "value": "Org1MSP::Admin@example.com",
  },
]
`;
